airbyte-webapp/src/services/connectorBuilder/ConnectorBuilderStateService.tsx

```diff
@@ -61,21 +60,23 @@ export const ConnectorBuilderFormStateProvider: React.FC<React.PropsWithChildren
   );
 
   const lastValidBuilderFormValuesRef = useRef<BuilderFormValues>(storedBuilderFormValues as BuilderFormValues);
+  const currentBuilderFormValuesRef = useRef<BuilderFormValues>(storedBuilderFormValues as BuilderFormValues);
 
   const setBuilderFormValues = useCallback(
     (values: BuilderFormValues, isValid: boolean) => {
       if (isValid) {
         // update ref first because calling setStoredBuilderFormValues might synchronously kick off a react render cycle.
         lastValidBuilderFormValuesRef.current = values;
       }
+      currentBuilderFormValuesRef.current = values;
       setStoredBuilderFormValues(values);
     },
     [setStoredBuilderFormValues]
   );
 
-  const builderFormValues = useMemo(() => {
-    return merge({}, DEFAULT_BUILDER_FORM_VALUES, storedBuilderFormValues);
-  }, [storedBuilderFormValues]);
+  // use the ref for the current builder form values because useLocalStorage will always serialize and deserialize the whole object,
+  // changing all the references which re-triggers all memoizations
+  const builderFormValues = currentBuilderFormValuesRef.current || DEFAULT_BUILDER_FORM_VALUES;
```

Comment:
Doing the merge on render is problematic because it will create a lot of new references, busting memoize barriers.